### ISSUE TITLE
Callback

### ISSUE THREAD

**Issue by das1969** [BobVul/GrowlToToast#4]
Hi

great work on this.

Should callback to the originating application be working?
Is this possible and therefore could this be included in future developments?

**Comment by BobVul** [comment]
Hmm. Does Growl itself support that? i.e. does that work on the other Displays?

If it does, then I'll have to look into it more... would be a fairly significant change, mostly because  it's a pain in the arse to handle activations from a Win10 notification on a standard desktop app. UWP would be easier but more restrictive in other ways.

**Comment by das1969** [comment]
Here's some details on callback

[http://www.growlforwindows.com/gfw/help/gntp.aspx#callbacks](url)

A while ago I got the c# source for GMail Notififer Plus from github, and amended it to send new mail notification growls. I got this working OK, and I could click on a growl to open the relevant email message in the browser.

I was tinkering with this again to get mail notifications in the Win10 action center. I  got it working to some extent using the [notifyicon class](https://msdn.microsoft.com/en-us/library/system.windows.forms.notifyicon%28v=vs.110%29.aspx) which creates a balloon notification. The I was using the balloontipclicked event, so that I could click on the action center message and the message opened in the browser. I'm not well-versed enough in c# to get this working properly though.

**Comment by BobVul** [comment]
Thanks for that link.

At the moment I'm implementing a non-visual display in Growl. Looks like callbacks are only available on visual displays, which might impose stricter timeouts.

Along with probably having to implement a COM server to be able to respond to a Win10 toast activation, and adapting my code for both bidirectional communication and keeping the toast handler open, this is not a very simple task, and would probably take a while. I do plan to try it when I have some time, but please don't be surprised if it happens to take a month or more.

Oh, do you happen to know any simple programs that send GNTP messages with callbacks? I don't actually use Growl (this was all on someone's request) so I've only been able to test with the built-in preview.

**Comment by das1969** [comment]
There's a list of growl integration libraries on the growl for windows [developers page](http://www.growlforwindows.com/gfw/developers.aspx).

The C# library I used was a Visual Studio solution containing examples of GNTP with callback.

**Comment by BobVul** [comment]
Ah... I suppose I could create a custom one but I was kinda hoping you had a real-world example.

Anyway, this'll do.

**Comment by mcormick** [comment]
Specifying an URL as /cu callback from growlnotify which invokes the callbackUrl does not seem to work in Toast version 2.0. Works with original "Displays". Any suggestions?

**Comment by BobVul** [comment]
@mcormick I don't believe this ever worked, or at least it was not intentionally implemented.

I might try as time permits, but the difficulty is the current system effectively shows a fire-and-forget notification, with no user interaction. I don't know how much the action center has changed since the first release but at the time that would have required a COM server to react to clicks.

Hmm. I wonder if turning Toaster into a UWP application would make it easier.

**Comment by BobVul** [comment]
The other half is communication from Growler (plugin) to Toaster is also one-way at the moment. That means even if Toaster were to receive click activations it would have no way to report back to Growl to trigger the callback.

Bidirectional communication would be enabled (made easier by?) #9.
